### PR TITLE
Unblock the controller's ability to register/deregister more than 100 targets

### PR DESCRIPTION
**Pull request by zijun726911**
**What type of PR is this?** feature/bugfix

### Changes:
- Unblock the controller's ability to register/deregister more than 100 targets

### Testing done on this change:

e2e test suite pass:
```
------------------------------

Ran 51 of 51 Specs in 2602.216 seconds
SUCCESS! -- 51 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestIntegration (2602.51s)
PASS
ok  	github.com/aws/aws-application-networking-k8s/test/suites/integration	2602.972s
```

Manual E2E test:
In my EKS cluster, deploy a `Deployment` with more than 100 replicas:
```
spec:
  replicas: 201
```

```
kubectl get deployment -A
NAMESPACE     NAME             READY     UP-TO-DATE   AVAILABLE   AGE
default       parking          201/201   201          201         30m
```
Setup all connectivity: gateway, httproute, k8sService. it could register 201 lattice targets:
image
image

Delete the Httproute, it could deregister all 201 targets and other lattice resources

----

Also did the` Deployment` replicas num mutation, it could change to correct num of lattice Targets. (lattice api register/deregister targets are fully idempotent, register an already registed target will return success) 

----

Also tested for ServiceExport, it could have more than 100 targets, changing the` Deployment` 
```
{"level":"info","ts":"2023-11-02T11:39:37.124-0700","logger":"controller.service-export","caller":"controllers/serviceexport_controller.go:106","msg":"reconcile","name":"parking"}
{"level":"info","ts":"2023-11-02T11:39:37.429-0700","logger":"controller.service-export","caller":"lattice/target_group_manager.go:133","msg":"Success UpdateTargetGroup tg-0561b02d07d3045a8"}
{"level":"info","ts":"2023-11-02T11:39:37.880-0700","logger":"controller.service-export","caller":"lattice/targets_manager.go:88","msg":"Success Register 100 Targets for partition(1/3) for target group: tg-0561b02d07d3045a8"}
{"level":"info","ts":"2023-11-02T11:39:38.206-0700","logger":"controller.service-export","caller":"lattice/targets_manager.go:88","msg":"Success Register 100 Targets for partition(2/3) for target group: tg-0561b02d07d3045a8"}
{"level":"info","ts":"2023-11-02T11:39:38.459-0700","logger":"controller.service-export","caller":"lattice/targets_manager.go:88","msg":"Success Register 1 Targets for partition(3/3) for target group: tg-0561b02d07d3045a8"}

```

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/deploy/lattice/target_group_manager.go [hunk 1]
```diff
@@ -195,22 +197,28 @@ func (s *defaultTargetGroupManager) Delete(ctx context.Context, modelTg *model.T
...
 
 	if len(targetsToDeregister) > 0 {
-		deRegisterInput := vpclattice.DeregisterTargetsInput{
-			TargetGroupIdentifier: &modelTg.Status.Id,
-			Targets:               targetsToDeregister,
-		}
+		var deregisterTargetsError error
+		partitions := getPartitionedLatticeTargets(targetsToDeregister, 100)
+		for i, targets := range partitions {
+			deRegisterInput := vpclattice.DeregisterTargetsInput{
```
Comment (solmonk) [inline pkg/deploy/lattice/target_group_manager.go:203]: nit: deRegister -> deregister, I know this has been like this before, but lets fix it when we have a chance.

Comment (xWink) [reply pkg/deploy/lattice/target_group_manager.go:203]: +1, same applies to other places like deRegResp -> deregResp

File: pkg/deploy/lattice/target_group_manager.go [hunk 2]
```diff
@@ -195,22 +197,28 @@ func (s *defaultTargetGroupManager) Delete(ctx context.Context, modelTg *model.T
...
+			isDeRegRespUnsuccessful := len(deRegResp.Unsuccessful) > 0
+			if isDeRegRespUnsuccessful {
+				deregisterTargetsError = errors.Join(deregisterTargetsError, fmt.Errorf("unsuccessful %d DeregisterTargets for partitions (%d/%d) for target group %s, deRegResp: %s",
+					len(deRegResp.Unsuccessful), i, len(partitions), modelTg.Status.Id, deRegResp))
 
-		deRegResp, err := lattice.DeregisterTargetsWithContext(ctx, &deRegisterInput)
-		if err != nil {
-			return fmt.Errorf("Failed DeregisterTargets %s due to %s", modelTg.Status.Id, err)
+			}
+			s.log.Infof("Success DeregisterTargets for partition (%d/%d) for target group %s", i+1, len(partitions), modelTg.Status.Id)
```
Comment (xWink) [inline pkg/deploy/lattice/target_group_manager.go:217]: This should be a debug log, don't need to see successful target deregistration on every partition in info mode

Comment (xWink) [reply pkg/deploy/lattice/target_group_manager.go:217]: Nit, clarity:
`Successfully deregistered targets to VPC Lattice Target Group %s for partition %d/%d`

File: pkg/deploy/lattice/target_group_manager.go [hunk 3]
```diff
@@ -195,22 +197,28 @@ func (s *defaultTargetGroupManager) Delete(ctx context.Context, modelTg *model.T
...
+				TargetGroupIdentifier: &modelTg.Status.Id,
+				Targets:               targets,
+			}
+			deRegResp, err := lattice.DeregisterTargetsWithContext(ctx, &deRegisterInput)
+			if err != nil {
+				return fmt.Errorf("Failed DeregisterTargets %s due to %s", modelTg.Status.Id, err)
+			}
+			isDeRegRespUnsuccessful := len(deRegResp.Unsuccessful) > 0
+			if isDeRegRespUnsuccessful {
+				deregisterTargetsError = errors.Join(deregisterTargetsError, fmt.Errorf("unsuccessful %d DeregisterTargets for partitions (%d/%d) for target group %s, deRegResp: %s",
```
Comment (xWink) [inline pkg/deploy/lattice/target_group_manager.go:213]: nit: clarity

`Failed to deregister targets from VPC Lattice Target Group %s for partition %d/%d due to %s`

File: pkg/deploy/lattice/targets_manager.go [hunk 4]
```diff
@@ -61,39 +62,33 @@ func (s *defaultTargetsManager) registerTargets(
...
+	partitionedLatticeTargets := getPartitionedLatticeTargets(latticeTargets, 100)
+	var registerTargetsError error
+	for i, targets := range partitionedLatticeTargets {
+		registerRouteInput := vpclattice.RegisterTargetsInput{
+			TargetGroupIdentifier: &modelTg.Status.Id,
+			Targets:               targets,
+		}
+		resp, err := s.cloud.Lattice().RegisterTargetsWithContext(ctx, &registerRouteInput)
+		if err != nil {
+			registerTargetsError = errors.Join(registerTargetsError, fmt.Errorf("failed RegisterTargets %s due to %s", modelTg.Status.Id, err))
```
Comment (xWink) [inline pkg/deploy/lattice/targets_manager.go:81]: Can use same clarity improvements from deregister targets errors

File: pkg/deploy/lattice/targets_manager.go [hunk 5]
```diff
@@ -111,23 +106,61 @@ func (s *defaultTargetsManager) deregisterStaleTargets(
...
 		if isStale {
 			targetsToDeregister = append(targetsToDeregister, &vpclattice.Target{Id: latticeTarget.Id, Port: latticeTarget.Port})
 		}
 	}
-
-	if len(targetsToDeregister) > 0 {
+	if len(targetsToDeregister) == 0 {
+		return nil
+	}
+	partitionedTargetsToDeregister := getPartitionedLatticeTargets(targetsToDeregister, 100)
```
Comment (mikhail-aws) [inline pkg/deploy/lattice/targets_manager.go:116]: Dont need getPartitionedLatticeTargets either

```go
	const chunkSize = 100
	for low := 0; low < len(targetsToDeregister); low += chunkSize {
		high := min(low+chunkSize, len(targetsToDeregister))
		chunk := targetsToDeregister[low:high]
	}
```

Unfortunately we still dont have min function in Go as of 1.20, but it's coming with 1.21 (any other sane language has it). You can add this one too.

```go
// built-in from golang 1.21 (we are at 1.20)
func min[T constraints.Ordered](a, b T) T {
	if a < b {
		return a
	}
	return b
}
```

**Review by xWink — COMMENTED** [review]
Overall looks good, would just like to see some logging/error message changes

File: pkg/deploy/lattice/targets_manager.go [hunk 6]
```diff
@@ -61,39 +62,33 @@ func (s *defaultTargetsManager) registerTargets(
...
-			Port: &port,
-		}
-		latticeTargets = append(latticeTargets, &t)
-	}
-
 	// No targets to register
-	if len(latticeTargets) == 0 {
+	if len(modelTargets.Spec.TargetList) == 0 {
 		return nil
 	}
```
Comment (mikhail-aws) [inline pkg/deploy/lattice/targets_manager.go:68]: redundant condition, loop below holds invariant

File: pkg/deploy/lattice/targets_manager.go [hunk 7]
```diff
@@ -61,39 +62,33 @@ func (s *defaultTargetsManager) registerTargets(
...
+	latticeTargets := modelTargetsToLatticeTargets(modelTargets.Spec.TargetList)
+	// Partition the targets into groups of 100, because that is the allowed max number of targets per RegisterTargets API call
+	// https://docs.aws.amazon.com/vpc-lattice/latest/APIReference/API_RegisterTargets.html
+	partitionedLatticeTargets := getPartitionedLatticeTargets(latticeTargets, 100)
```
Comment (mikhail-aws) [inline pkg/deploy/lattice/targets_manager.go:72]: 100 should be const

File: pkg/deploy/lattice/targets_manager.go [hunk 8]
```diff
@@ -61,39 +62,33 @@ func (s *defaultTargetsManager) registerTargets(
...
+			TargetGroupIdentifier: &modelTg.Status.Id,
+			Targets:               targets,
+		}
+		resp, err := s.cloud.Lattice().RegisterTargetsWithContext(ctx, &registerRouteInput)
+		if err != nil {
+			registerTargetsError = errors.Join(registerTargetsError, fmt.Errorf("failed RegisterTargets %s due to %s", modelTg.Status.Id, err))
+		}
+		if len(resp.Unsuccessful) > 0 {
+			registerTargetsError = errors.Join(registerTargetsError, fmt.Errorf("failed RegisterTargets (Unsuccessful=%d) %s, will retry",
+				len(resp.Unsuccessful), modelTg.Status.Id))
```
Comment (mikhail-aws) [inline pkg/deploy/lattice/targets_manager.go:85]: Should we add each unsuccessful target into error? Dont see harm in it.
Why you mention "will retry" but there is not retry in this code. Also we should retry only on unsuccessful targets. Are we retrying full reconciliation loop in case there is 1 unsuccessful registration?

**Review by mikhail-aws — CHANGES_REQUESTED** [review]

File: pkg/deploy/lattice/targets_manager.go (hunk 6 above)
Comment (zijun726911) [reply pkg/deploy/lattice/targets_manager.go:68]: Return earlier to reduce cognitive load?

File: pkg/deploy/lattice/targets_manager.go (hunk 5 above)
Comment (zijun726911) [reply pkg/deploy/lattice/targets_manager.go:116]: I could change to this style, but still think put Partition logic in a separate function make it clearer

File: pkg/deploy/lattice/targets_manager.go [hunk 9]
```diff
@@ -111,23 +106,61 @@ func (s *defaultTargetsManager) deregisterStaleTargets(
...
+		port := modelTarget.Port
+		targetIP := modelTarget.TargetIP
+		t := vpclattice.Target{
+			Id:   &targetIP,
+			Port: &port,
+		}
+		targets = append(targets, &t)
+	}
+	return targets
 }
```
[the remark this answers is not in the document]
Comment (zijun726911) [reply pkg/deploy/lattice/targets_manager.go:166]: Thanks, will change

File: pkg/deploy/lattice/targets_manager.go (hunk 8 above)
Comment (zijun726911) [reply pkg/deploy/lattice/targets_manager.go:85]: Callers' or this function will do LATTICE_RETRY, I can remove "will retry" in this error message. 


> Also we should retry only on unsuccessful targets. Are we retrying full reconciliation loop in case there is 1 unsuccessful registration?

Current controller don't have a place to store unsuccessful registration(if we have a place to easily store unsuccessful targets we could retry only for unsuccessful for sure). Retry to register all targets is safe, as the vpc lattice registerTargets api is idempotent

File: pkg/deploy/lattice/targets_manager.go (hunk 6 above)
Comment (mikhail-aws) [reply pkg/deploy/lattice/targets_manager.go:68]: Less code reduces cognitive load, de-duplicating code also reduces cognitive load. Redundant checks and validations create fuzz around important code.

File: pkg/deploy/lattice/targets_manager.go (hunk 8 above)
Comment (mikhail-aws) [reply pkg/deploy/lattice/targets_manager.go:85]: Please remove "retry", code that actually performs retry should log/wrap err. The full Reconcile loop seems overkill... but out of scope.

Comment (solmonk) [reply pkg/deploy/lattice/targets_manager.go:85]: The current justification for requiring full reconcile loop is that we must go through routes to figure out which target groups are affected by such endpoint changes.
There's room for improvement for sure, but might need a big change for this.

File: pkg/deploy/lattice/targets_manager.go (hunk 6 above)
Comment (zijun726911) [reply pkg/deploy/lattice/targets_manager.go]: Removed

File: pkg/utils/common_test.go [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+		got := Chunks(in, size)
+		assert.Equal(t, [][]*struct{}{{obj1, obj2}}, got)
+
+	})
+
+	t.Run("[obj1, obj2, obj3] -> [[obj1,obj2],[obj3]]", func(t *testing.T) {
+		in := []*struct{}{obj1, obj2, obj3}
+		got := Chunks(in, size)
+		assert.Equal(t, [][]*struct{}{{obj1, obj2}, {obj3}}, got)
+	})
```
Comment (mikhail-aws) [inline pkg/utils/common_test.go:35]: Usually these functions tested well with table-style tests. Also sub-tests cover edge cases and normal cases. [obj1] -> [[obj1]] weird name of single item. Need to cover: no-items, single item, multiple items but single partition, mutliple items 2 partitions, multiple items 2+ partitions, partition size of 0(?)

**Review by mikhail-aws — APPROVED** [review]

File: pkg/utils/common_test.go [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+	"github.com/stretchr/testify/assert"
+)
+
+func TestChunks(t *testing.T) {
+	size := 2
+	obj1 := &struct{}{}
+	obj2 := &struct{}{}
+	obj3 := &struct{}{}
+
+	t.Run("[obj1] -> [[obj1]]", func(t *testing.T) {
```
Comment (mikhail-aws) [inline pkg/utils/common_test.go]: please use descriptive test names, "single item - single partition"

**Review by xWink — APPROVED** [review]

File: pkg/utils/common_test.go [hunk 12]
```diff
@@ -0,0 +1,78 @@
+package utils
+
+import (
+	"reflect"
+	"testing"
+)
+
+func TestChunks(t *testing.T) {
```
Comment (mikhail-aws) [inline pkg/utils/common_test.go:8]: I would do it differently... struct{}, hand-made slices, cryptic names makes it harder than it should be.

```go
import (
	"fmt"
	"testing"
)

func TestChunks(t *testing.T) {

	type test struct {
		sliceLen         int
		chunkSize        int
		wantChunksLen    int
		wantLastChunkLen int
	}

	tests := []test{
		{0, 1, 0, 0},
		{1, 1, 1, 1},
		{2, 1, 2, 1},
		{10, 2, 5, 2},
		{102, 10, 11, 2},
	}

	for _, tt := range tests {
		t.Run(fmt.Sprintf("sliceLen=%d, chunkSize=%d", tt.sliceLen, tt.chunkSize), func(t *testing.T) {
			slice := makeIntSlice(tt.sliceLen)
			chunks := Chunks(slice, tt.chunkSize)
			gotChunksLen := len(chunks)
			if gotChunksLen != tt.wantChunksLen {
				t.Errorf("number of chunks does not match, want=%d, got=%d", tt.wantChunksLen, gotChunksLen)
			}
			if gotChunksLen > 0 {
				lastChunk := chunks[gotChunksLen-1]
				if len(lastChunk) != tt.wantLastChunkLen {
					t.Errorf("last chunk size does not match, want=%d, got=%d", tt.wantLastChunkLen, len(lastChunk))
				}
			}
		})
	}
}

func makeIntSlice(s int) []int {
	out := make([]int, s)
	for i := 0; i < s; i++ {
		out[i] = i
	}
	return out
}
```

Comment (zijun726911) [reply pkg/utils/common_test.go:8]: Ok, looks good, strong unit tests 👍

File: pkg/utils/common.go [hunk 13]
```diff
@@ -4,12 +4,32 @@ import (
...
+	out := [][]T{}
+	if size <= 0 {
+		return out
+	}
+	for low := 0; low < len(in); low += size {
+		high := min(low+size, len(in))
+		out = append(out, in[low:high])
+	}
+	return out
+}
```
Comment (mikhail-aws) [inline pkg/utils/common.go:31]: Also a rule of thumb, especially for generic algorithms, keep them performant, with low overhead.
For example your and mine version use append. Append - will copy underlying items every time when underlying array cannot hold items anymore, so it will allocate new memory, twice bigger than before and copy all items.

So if you know exact size of the slice you want, it's better to preallocate memory and then use indices. It's not a big deal at all for current usage, but it makes difference when you work with thousands of items in the list. Especially not pointer values, but structs that are not cheap to copy.

```go
func Chunks[T any](in []T, size int) [][]T {
	out := make([][]T, len(in)/size+1)
	if size <= 0 {
		return out
	}
	for low := 0; low < len(in); low += size {
		high := min(low+size, len(in))
                chunkId := low / size
		out[chunkId] = in[low:high]
	}
	return out
}
```
